include/unifex/async_scope.hpp

```diff
@@ -632,76 +625,129 @@ class _attached_op<Sender, Receiver>::type final {
+    return reinterpret_cast<async_scope*>(scope & mask);
+  }
+
+  template <typename Func>
+  void deliver_result(Func func) noexcept {
+    auto scope = deliver_result_prelude();
+    if (scope) {
+      func(std::move(receiver_));
+      record_done(scope);
+    }
+  }
+
+  auto get_token() noexcept { return stopSource_.get_token(); }
+
+  const auto& get_receiver() const noexcept { return receiver_; }
 
   void deregister_callbacks() noexcept {
-    stopSource_.deregister_callbacks();
+    receiverCallback_.destruct();
+    scopeCallback_.destruct();
+  }
+
+  void register_callbacks() noexcept {
+    receiverCallback_.construct(
+        get_stop_token(receiver_), stop_callback{*this});
+    scopeCallback_.construct(
+        scope_ref()->get_stop_token(), stop_callback{*this});
   }
 
   friend void tag_invoke(tag_t<start>, type& op) noexcept {
     if (op.scope_ref()) {
-      op.set_started();
-      op.stopSource_.register_callbacks(
-          op.receiverToken_, op.scope_ref()->get_stop_token());
+      // TODO: callback constructor may throw
+      // 1. catch and propagate with `set_error`
+      // 2. careful when destroying callbacks (0, 1, 2?)
+      op.register_callbacks();
       unifex::start(op.op_.get());
     } else {
-      unifex::set_done(std::move(op).receiver_.get());
+      unifex::set_done(std::move(op).receiver_);
     }
   }
 
-  async_scope* scope_ref() const noexcept {
-    return reinterpret_cast<async_scope*>(scope_ & started_mask);
-  }
-
 private:
-  static constexpr std::uintptr_t started_bit{1};
-  static constexpr std::uintptr_t started_mask = ~started_bit;
-  // async_scope* stored as integer, so LSB can be used to signal start
-  std::uintptr_t scope_;
+  static constexpr std::uintptr_t mask = ~(1u | std::uintptr_t{1u << 1});
+  // async_scope* stored as integer: low 2 bits for ref count
+  std::atomic_uintptr_t scope_;
   using receiver_stop_token_t = stop_token_type_t<Receiver>;
   using scope_stop_token_t = inplace_stop_token;
-  fused_stop_source<receiver_stop_token_t, scope_stop_token_t> stopSource_;
-  receiver_stop_token_t receiverToken_;
-  union {
-    manual_lifetime<nested_operation_t> op_;
-    manual_lifetime<Receiver> receiver_;
-  };
+  template <typename StopToken>
+  using stop_callback_t = manual_lifetime<
+      typename StopToken::template callback_type<stop_callback>>;
+  UNIFEX_NO_UNIQUE_ADDRESS inplace_stop_source stopSource_;
+  UNIFEX_NO_UNIQUE_ADDRESS stop_callback_t<receiver_stop_token_t>
+      receiverCallback_;
+  UNIFEX_NO_UNIQUE_ADDRESS stop_callback_t<scope_stop_token_t> scopeCallback_;
+  UNIFEX_NO_UNIQUE_ADDRESS Receiver receiver_;
+  UNIFEX_NO_UNIQUE_ADDRESS manual_lifetime<nested_operation_t> op_;
 
-  bool has_started() const noexcept { return scope_ & started_bit; }
-  void set_started() noexcept { scope_ |= started_bit; }
+  async_scope* scope_ref() const noexcept {
+    return reinterpret_cast<async_scope*>(
+        scope_.load(std::memory_order_relaxed) & mask);
```

Comment:
With the static helpers, this could be:
```suggestion
    return scope_ptr(scope_.load(std::memory_order_relaxed));
```